pkg/volume/csi/csi_attacher.go

```diff
@@ -106,22 +107,33 @@ func (c *csiAttacher) Attach(spec *volume.Spec, nodeName types.NodeName) (string
 		},
 	}
 
-	_, err = c.k8s.StorageV1().VolumeAttachments().Create(context.TODO(), attachment, metav1.CreateOptions{})
-	alreadyExist := false
+	// Begin watching for this volume attachment prior to creating it to avoid race conditions.
+	if _, err = c.plugin.volumeAttachmentInformerHook.CreateWatch(attachID, watchChannelBufferSize); err != nil {
```

Comment:
I am not very sure about watch design.

Previously, a watch is created for every unique volumeId. For attach/detach operation, we do not allow concurrent operations. So for each unique volumeID, only one attach or detach operation is allowed. So the total number of watch is equal to the total number of unique volumes. 

For your new design, the number of watch will still be the total number of unique volumes?

Reply:
Yes. If we're already watching for a particular VolumeAttachment then this `CreateWatch()` call will fail.

Reply:
then what benefit we can get from this?

Reply:
To clarify, there will be one API watch for VolumeAttachments, which is the one used by the SharedInformer. We will be "watching for" numerous VolumeAttachments by dynamically hooking into this SharedInformer, but under the hood this is not creating more API watches.

Reply:
oh, I think I got it now. Instead of calling API server to create each watch, you use sharedInformer to create those watch.

Reply:
The "watch" that we create via this SharedInformerHook is not the same as the "watch" that we create via the API. The "watch" that we create via the SharedInformerHook is just adding an entry to a map of volume attachment IDs that we will keep track of.